File: YEAR 2/SEM 1/Advanced Programming Methods/LAB 4/src/Repository/MemoryRepository.java
package Repository;

import Model.State.ProgramState;
import Model.Statement.IStatement;
import Util.Dictionary.MyIDictionary;
import Util.List.MyIList;
import Util.Stack.MyIStack;

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.List;

public class MemoryRepository implements IRepository {
    private List<ProgramState> list;
    private String logFilePath;

    public MemoryRepository() {
        this.logFilePath = null;
        this.list = new ArrayList<>();
    }

    public MemoryRepository(String logFilePath) {
        this.logFilePath = logFilePath;
        this.list = new ArrayList<>();
    }

    private void logState(ProgramState state,PrintWriter logWriter) {
        MyIStack<IStatement> executionStack = state.getExecutionStack();
        MyIDictionary<String,Integer> symbolTable = state.getSymbolTable();
        MyIList<Integer> output = state.getOutput();

        logWriter.println("STACK:");
        logWriter.println(executionStack);
        logWriter.println("SYMBOL TABLE:");
        logWriter.println(symbolTable);
        logWriter.println("OUTPUT:");
        logWriter.println(output);
        logWriter.println("FILE TABLE:");
        logWriter.println(state.getFileTable());
        logWriter.println("HEAP:");
        logWriter.println(state.getHeap());
    }

    @Override
    public void add(ProgramState element) {
        this.list.add(element);
    }

    @Override
    public ProgramState back() throws RepositoryException {
        if(this.list.size() == 0)
            throw new RepositoryException("Empty repository!");

        return this.list.get(this.list.size() - 1);
    }

    @Override
    public void logData() throws RepositoryException {
        PrintWriter logWriter;

        try {
            logWriter = new PrintWriter(new BufferedWriter(new FileWriter(this.logFilePath, true)));
        }catch(IOException ex) {
            throw new RepositoryException("Could not open log file!");
        }

        for(ProgramState state : this.list) {
            logWriter.println("========== STATE ==========");
            this.logState(state,logWriter);
        }

        logWriter.println("---------------------------------");
        logWriter.close();
    }
}

File: YEAR 2/SEM 1/Advanced Programming Methods/ToyLanguage/ToyLanguage/Repository/MemoryRepository.cs
using System;
using ToyLanguage.Model.State;
using ToyLanguage.Util.List;
using System.IO;

namespace ToyLanguage.Repository
{
    class MemoryRepository : IRepository
    {
        private readonly MyIList<ProgramState> data = new MyList<ProgramState>();
        private readonly string logFile;

        public MemoryRepository(string logFile)
        {
            if (File.Exists(logFile))
                File.Delete(logFile);

            this.logFile = logFile;
        }

        public bool IsEmpty => data.Count == 0;

        public void Add(ProgramState programState)
        {
            data.Add(programState);
        }

        public ProgramState Get(int index)
        {
            if (index < 0 || index > data.Count) throw new MemoryRepositoryException("Invalid index!");

            return data.ElementAt(index);
        }
        
        private void LogProgramState(ProgramState programState,StreamWriter writer)
        {
            writer.WriteLine("==== Program State ====");
            writer.WriteLine("ExecutionStack:");
            writer.WriteLine(programState.ExecutionStack);
            writer.WriteLine("SymbolTable:");
            writer.WriteLine(programState.SymbolTable);
            writer.WriteLine("Output:");
            writer.WriteLine(programState.Output);
            writer.WriteLine("File table:");
            writer.WriteLine(programState.FileTable);
        }

        public void LogData()
        {
            try
            { 
                StreamWriter writer = new StreamWriter(logFile, true);
            
                for(int i = 0; i < data.Count; i++)
                {
                    LogProgramState(data.ElementAt(0),writer);
                }
                
                writer.Close();
            }catch(Exception)
            {
                throw new MemoryRepositoryException("Could not write data to log file!");
            }
        }
    }
}
